src/codegen/compile.ml

```diff
@@ -5642,6 +5643,35 @@ module MakeSerialization (Strm : Stream) = struct
       (fun env -> compile_unboxed_const 0l) ^^
     deserialize_from_blob false env ts
 
+(*
+Note [speculating for short (S)LEB encoded bignums]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#3098 highlighted that a lot of heap garbage can be generated while reading in
+(S)LEB-encoded bignums. To make heap consumption optimal for every compactly
+representable `Int`, we resort to speculatively reading a 64-byte chunk from
+the `ReadBuf`. We call it speculative, because it may read past the end of the
+buffer (and thus end up containing junk bytes) or even fail because reading
+across Wasm page boundaries could cause segfaulting. (Consider the buffer ending
```

Comment:
I don't think wasm can segfault. Do you mean trap?

Also I didn't think reading across page boundaries is a problem. Do you mean reading across the end of the memory?

Reply:
It can  trap when we read past the last _allocated_ page's boundary. Since we don't want to get the number of allocated pages here, we just exclude the last few bytes of _each_  page from consideration.

Reply:
But would that trap be bad? Shouldn't we trap eventually anyway if the leb encoding runs off end off memory?

Reply:
Fixed the wording. Also the `(Consider the buffer ending  ...` remark does describe the problem at hand.

Reply:
@crusso The leb-encoded buffer could be just 1-2 bytes, and we want to read 8. The read could clearly run into nowhere-land, but the buffer still ends cleanly in an allocated page.

Reply:
Ah OK.

But actually, why did you discard the car solutiin above? Just gave two loops, the first reading up to five bytes one by one into a word64, returning if needed, followed by a second loop reading the remaining bytes into a bignum initialized  from the word64?

Reply:
I never implemented 🚘. Going via Rust would give an inferior cycle count and inlining the loop would bloat the Wasm each time  `Int/Nat` appears in a message (or stable `var`).

Reply:
@crusso exploring this speculative method will come in handy when speeding up `bigint_(s)leb128_decode` which does a `read_byte` in a loop now.

Reply:
> I never implemented 🚘. Going via Rust would give an inferior cycle count and inlining the loop would bloat the Wasm each time `Int/Nat` appears in a message (or stable `var`).

Don't we always ultimately go via rust anyway? I think isolating the changes to the original Rust routines, adding a < 5 iteration loop, would have been simpler and easier to grok, even if slightly less efficient with cycles.